Create proc/func not visible in the pg_stat_statements view

kritika-0601 opened the pull request:
### Description

Currently Babelfish does not show create proc/func query inside the pg_stat_statements table. After my change now it is visible in the view.
pg_stat_statements displays planning and execution statistics for all SQL statements executed by a server. Some statements like create proc/func was not being tracked. Mainly because subsequent processutility hooks after bbf_processutilty were not being called.
Created a hook for that and called it inside relevant ProcessUtility function.



[Describe what this change achieves]
 

### Issues Resolved

[List any issues this PR will resolve]
 
### Check List

- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV__PG_15_X' into pg_stat_statements-kritika

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
retrigger tests

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV__PG_15_X' into pg_stat_statements-kritika

Commit:
retrigger tests

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

rishabhtanwar29 commented on src/backend/tcop/utility.c:
```diff
@@ -1656,8 +1662,44 @@ ProcessUtilitySlow(ParseState *pstate,
+					}
+
+					if (language_item)
+						language = strVal(language_item->arg);
```
I think we can minimize the engine changes by moving all this logic to extension itself since we are anyway adding a hook. Here we can directly call the hook if it exists, something like:
```
                    if (CreateFunctionStmt_hook)
					{
						address = (*CreateFunctionStmt_hook)(pstate, pstmt, queryString, false, context, params, queryEnv, dest, qc);
					}
					else
					{
						address = CreateFunction(pstate, (CreateFunctionStmt *) parsetree);
					}
```
The required `language` or `dialect` checks can be done in the hook itself.

rishabhtanwar29 commented on src/backend/tcop/utility.c:
```diff
@@ -73,9 +73,15 @@
+#include "parser/parse_type.h"
+
+#include "tcop/cmdtag.h"
+#include "tcop/tcopprot.h"
```
Let's move these above with the other includes in alphabetical order. Also, make sure to verify if we really need all of them.

rishabhtanwar29 commented on src/include/parser/parse_type.h:
```diff
@@ -15,6 +15,13 @@
+#include "tcop/cmdtag.h"
+#include "parser/parse_type.h"
+#include "nodes/plannodes.h"
+#include "nodes/params.h"
```
Let's order them alphabetically.

rishabhtanwar29 commented on src/include/parser/parse_type.h:
```diff
@@ -68,4 +75,16 @@ extern PGDLLIMPORT check_or_set_default_typmod_hook_type check_or_set_default_ty
 typedef void (*validate_var_datatype_scale_hook_type)(const TypeName *typeName, Type typ);
 extern PGDLLIMPORT validate_var_datatype_scale_hook_type validate_var_datatype_scale_hook;
 
+typedef void (*CreateFunctionStmt_hook_type)(ParseState *pstate,
```
I think better header file to add this hook would be `defrem.h` or `utility.h`

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

forestkeeper commented on src/backend/tcop/utility.c:
```diff
@@ -1656,8 +1658,15 @@ ProcessUtilitySlow(ParseState *pstate,
+						address = CreateFunction(pstate, (CreateFunctionStmt *) parsetree);
+					}
+					break;
+			}
```
this } doesn't match space with {

forestkeeper commented on src/include/parser/parse_type.h:
```diff
@@ -15,6 +15,8 @@
 
 #include "access/htup.h"
 #include "parser/parse_node.h"
+#include "tcop/cmdtag.h"
```
why add those two .h ?

forestkeeper commented on src/backend/tcop/utility.c:
```diff
@@ -61,6 +61,7 @@
 #include "commands/view.h"
 #include "miscadmin.h"
 #include "parser/parser.h"
+#include "parser/parse_type.h"
```
should add utility.h not parse_type.h in here

Commit:
Merge branch 'BABEL_3_X_DEV__PG_15_X' into pg_stat_statements-kritika

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

Commit:
Create proc/func not visible in the pg_stat_statements view

Subsequent processutility hooks after bbf_processutilty were not being
called. Created a hook for that and called it inside relevant
ProcessUtility function.

Signed-off-by: Kritika <kritikaq@amazon.com>

kritika-0601 committed:
Merge branch 'babelfish-for-postgresql:BABEL_3_X_DEV__PG_15_X' into pg_stat_statements-kritika
